File: src/main/java/cz/cvut/fel/plichjan/distmesh/inputs/DRectangle.java
package cz.cvut.fel.plichjan.distmesh.inputs;

/**
 * function d=drectangle(p,x1,x2,y1,y2)
 * d=-min(min(min(-y1+p(:,2),y2-p(:,2)),-x1+p(:,1)),x2-p(:,1));
 */
public class DRectangle implements IDistanceFunction {
    private double x1,x2,y1,y2;

    //function d=drectangle(p,x1,x2,y1,y2)
    public DRectangle(double x1, double x2, double y1, double y2) {
        this.x1 = x1;
        this.x2 = x2;
        this.y1 = y1;
        this.y2 = y2;
    }

    @Override
    public double call(double x, double y) {
        // d=-min(min(min(-y1+p(:,2),y2-p(:,2)),-x1+p(:,1)),x2-p(:,1));
        return -Math.min(Math.min(Math.min(-y1 + y, y2 - y), -x1 + x), x2 - x);
    }
}

File: src/main/kotlin/cz/cvut/fel/plichjan/distmesh/inputs/DRectangle.kt
package cz.cvut.fel.plichjan.distmesh.inputs

import kotlin.math.min

/**
 * function d=drectangle(p,x1,x2,y1,y2)
 * d=-min(min(min(-y1+p(:,2),y2-p(:,2)),-x1+p(:,1)),x2-p(:,1));
 */
data class DRectangle(val x1: Double, val x2: Double, val y1: Double, val y2: Double) : IDistanceFunction {
    override fun call(x: Double, y: Double): Double {
        return -min(min(min(-y1 + y, y2 - y), -x1 + x), x2 - x)
    }
}
